pkg/proxy/iptables/proxier_test.go

```diff
@@ -1111,6 +1113,258 @@ COMMIT
 	}
 }
 
+// cidrContains checks if cidrStr contains ipStr. cidrStr can actually be a CIDR or an IP
+func cidrContains(cidrStr, ipStr string) bool {
+	ip := netutils.ParseIPSloppy(ipStr)
+	if ip == nil {
+		panic("bad IP in test case: " + ipStr)
+	}
+
+	if strings.Contains(cidrStr, "/") {
+		_, cidr, err := netutils.ParseCIDRSloppy(cidrStr)
+		if err != nil {
+			panic("bad CIDR in kube-proxy output: " + err.Error())
+		}
+		return cidr.Contains(ip)
+	} else {
+		ip2 := netutils.ParseIPSloppy(cidrStr)
+		if ip2 == nil {
+			panic("bad IP in kube-proxy output: " + cidrStr)
+		}
+		return ip.Equal(ip2)
+	}
+}
+
+var commentRegex = regexp.MustCompile(`-m comment --comment ("[^"]*"|[^" ]*) `)
+var moduleRegex = regexp.MustCompile("-m ([^ ]*)")
+var localRegex = regexp.MustCompile("(!)? --(src|dst)-type LOCAL")
+var destIPRegex = regexp.MustCompile("(!)? -d ([^ ]*) ")
+var destPortRegex = regexp.MustCompile(" --dport ([^ ]*) ")
+var sourceIPRegex = regexp.MustCompile("(!)? -s ([^ ]*) ")
+var affinityRegex = regexp.MustCompile(" --rcheck ")
+var probabilityRegex = regexp.MustCompile(" --probability ([^ ]*) ")
+var jumpToEndpointRegex = regexp.MustCompile(" -j KUBE-SEP-")
+
+var noMatchModules = sets.NewString("conntrack", "mark")
+var knownModules = sets.NewString("addrtype", "recent", "statistic", "tcp")
+
+// ruleMatches checks if the give iptables rule matches a packet with the given sourceIP,
+// destIP, and destPort. nodeIP is the local node IP (for rules matching "LOCAL").
+// endpointProbability is the probability string to match for an endpoint jump (or "" to
+// match the final, fallback jump)
+func ruleMatches(rule, sourceIP, destIP, destPort, nodeIP, endpointProbability string) bool {
+	var match []string
+
+	rule = commentRegex.ReplaceAllString(rule, "")
+
+	for _, moduleMatches := range moduleRegex.FindAllStringSubmatch(rule, -1) {
+		if noMatchModules.Has(moduleMatches[1]) {
+			return false
+		} else if !knownModules.Has(moduleMatches[1]) {
+			klog.ErrorS(nil, "Rule uses unknown iptables module; assuming no match", "rule", rule, "module", moduleMatches[1])
+			return false
+		}
+	}
+
+	// The sub-rules within an iptables rule are ANDed together, so the rule only
+	// matches if all of them match. So go through the subrules, and if any of them
+	// don't match, then fail.
+
+	// Match local/non-local
+	match = localRegex.FindStringSubmatch(rule)
+	if match != nil {
+		sourceIsLocal := sourceIP == nodeIP || sourceIP == "127.0.0.1"
```

Comment:
is `::1` required?

Reply:
No, nodeports (and NAT in general) don't work with `::1`.

I think what actually happens is that the rule matches but then the `-j DNAT` fails to actually change the destination address. But pretending the rule doesn't match has the same effect.

Reply:
(also, we don't use IPv6 in any of the current tests. "not fully generic")